pkg/kstatus/observe/observer/observer.go

```diff
@@ -0,0 +1,197 @@
+package observer
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/observe/event"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// AggregatorFactoryFunc defines the signature for the function the Observer will use to
+// create a new StatusAggregator for each StatusObserverRunner.
+type AggregatorFactoryFunc func(identifiers []wait.ResourceIdentifier) StatusAggregator
+
+// ReaderFactoryFunc defines the signature for the function the Observer will use to create
+// a new ObserverReader for each StatusObserverRunner.
+type ReaderFactoryFunc func(reader client.Reader, mapper meta.RESTMapper, identifiers []wait.ResourceIdentifier) (ObserverReader, error)
+
+// ObserversFactoryFunc defines the signature for the function the Observer will use to
+// create the resource observers and the default observer for each StatusObserverRunner.
+type ObserversFactoryFunc func(reader ObserverReader, mapper meta.RESTMapper) (resourceObservers map[schema.GroupKind]ResourceObserver, defaultObserver ResourceObserver)
+
+// Observer provides functionality for polling a cluster for status for a set of resources.
+type Observer struct {
+	Reader client.Reader
+	Mapper meta.RESTMapper
+
+	AggregatorFactoryFunc AggregatorFactoryFunc
+	ReaderFactoryFunc     ReaderFactoryFunc
+	ObserversFactoryFunc  ObserversFactoryFunc
+}
+
+// Observe will create a new StatusObserverRunner that will poll all the resources provided and report their status
+// back on the event channel returned. The StatusObserverRunner can be cancelled at any time by cancelling the
+// context passed in.
+// If stopOnCompleted is set to true, then the runner will stop observing the resources when the StatusAggregator
+// determines that all resources has been fully reconciled. If this is set to false, the observer will keep running
+// until cancelled.
+// The pollInterval specifies how often the observer should poll the cluster for the latest state of the resources.
+func (s *Observer) Observe(ctx context.Context, identifiers []wait.ResourceIdentifier, pollInterval time.Duration, stopOnCompleted bool) <-chan event.Event {
```

Comment:
linefeeds would be nice for reading and reviewing on a small screen (e.g laptop).
fwiw, i follow the cstyle policy: http://go/cstyle#Line_Length, but don't require others too
(just hope they will)